specs/core/1_shard-data-chains.md

```diff
@@ -299,78 +331,201 @@ def compute_crosslink_data_root(blocks: Sequence[ShardBlock]) -> Hash:
-        assert block.attester_bitfield[i] is False
+            add_reward(state, beacon_state, index, base_reward)
+            attestations += 1
+
+    for i in range(len(attester_committee), MAX_PERSISTENT_COMMITTEE_SIZE):
+        assert block.core.attester_bitfield[i] is False or block.core.attester_bitfield[i] == 0  # TODO: FIX Bitvector
+
     assert bls_verify(
         pubkey=bls_aggregate_pubkeys(pubkeys),
-        message_hash=candidate.core.parent_root,
-        signature=candidate.signatures.attestation_signature,
-        domain=get_domain(beacon_state, DOMAIN_SHARD_ATTESTER, compute_epoch_of_shard_slot(candidate.core.slot))
+        message_hash=block.core.parent_root,
+        signature=block.signatures.attestation_signature,
+        domain=get_domain(beacon_state, DOMAIN_SHARD_ATTESTER, compute_epoch_of_shard_slot(block.core.slot))
     )
 
     # Check proposer
-    proposer_index = get_shard_block_proposer_index(beacon_state, shard, candidate.core.slot)
+    proposer_index = get_shard_block_proposer_index(beacon_state, state.shard, block.core.slot)
     assert proposer_index is not None
+    add_reward(state, beacon_state, proposer_index, Gwei(attestations * base_reward // PROPOSER_REWARD_QUOTIENT))
     assert bls_verify(
         pubkey=beacon_state.validators[proposer_index].pubkey,
-        message_hash=hash_tree_root(candidate.core),
-        signature=candidate.signatures.proposer_signature,
-        domain=get_domain(beacon_state, DOMAIN_SHARD_PROPOSER, compute_epoch_of_shard_slot(candidate.core.slot)),
+        message_hash=hash_tree_root(block.core),
+        signature=block.signatures.proposer_signature,
+        domain=get_domain(beacon_state, DOMAIN_SHARD_PROPOSER, compute_epoch_of_shard_slot(block.core.slot)),
     )
 
-    return True
+    # Process and update block data fees
+    add_fee(state, beacon_state, proposer_index, Gwei(state.basefee * len(block.core.data) // SHARD_BLOCK_SIZE_LIMIT))
+    QUOTIENT = SHARD_BLOCK_SIZE_LIMIT * BASEFEE_ADJUSTMENT_FACTOR
+    if len(block.core.data) > SHARD_BLOCK_SIZE_TARGET:
+        state.basefee += Gwei(min(1, state.basefee * (len(block.core.data) - SHARD_BLOCK_SIZE_TARGET) // QUOTIENT))
+    elif len(block.core.data) < SHARD_BLOCK_SIZE_TARGET:
+        state.basefee -= Gwei(min(1, state.basefee * (len(block.core.data) - SHARD_BLOCK_SIZE_TARGET) // QUOTIENT))
+    state.basefee = Gwei(max(
+        1,
+        min(
+            EFFECTIVE_BALANCE_INCREMENT // EPOCHS_PER_SHARD_PERIOD // SHARD_SLOTS_PER_BEACON_SLOT * SLOTS_PER_EPOCH,
+            state.basefee,
+        )
+    ))
+
+    # Check total bytes
+    assert block.core.total_bytes == state.most_recent_block_core.total_bytes + len(block.core.data)
+
+    # Update in-state block header
+    state.most_recent_block_core = ShardBlockCore(
+        slot=block.core.slot,
+        beacon_chain_root=block.core.beacon_chain_root,
+        parent_root=block.core.parent_root,
+        data_root=block.core.data_root,
```

Comment:
`block.core` is `ExtendedShardBlockCore`, so it looks that we need a tree hashing here?

```suggestion
        data_root=hash_tree_root(block.core.data),
```